conmon-rs/server/src/cri_logger.rs

```diff
@@ -0,0 +1,256 @@
+            file: Self::open(&path).await?,
+            max_log_size,
+        })
+    }
+
+    /// Write the contents of the provided reader into the file logger. Ensure to manually call
+    /// `flush` if you need the written log data on disk.
+    pub async fn write<T>(&mut self, pipe: Pipe, reader: &mut BufReader<T>) -> Result<()>
+    where
+        T: AsyncBufRead + Unpin,
+    {
+        // Get the RFC3339 timestmap
+        let timestamp = Local::now().to_rfc3339();
+        let min_log_len = timestamp
+            .len()
+            .checked_add(10) // len of " stdout " + "P "
+            .context("min log line len exceeds usize")?;
+        let mut bytes_written = 0;
+
+        loop {
+            // Read the line
+            let mut line_buf = Vec::with_capacity(min_log_len);
+            let (read, partial) = Self::read_line(reader, &mut line_buf).await?;
+
+            if read == 0 {
+                break;
+            }
+
+            let mut bytes_to_be_written = read + min_log_len;
+            if partial {
+                bytes_to_be_written += 1; // the added newline
+            }
+
+            if let Some(max_log_size) = self.max_log_size() {
+                trace!(
+                    "Verifying log size: max_log_size = {}, bytes_written = {}, bytes_to_be_written = {}", 
+                    max_log_size, bytes_written, bytes_to_be_written,
+                );
+                if (bytes_written + bytes_to_be_written) > max_log_size {
+                    bytes_written = 0;
+                    self.reopen()
+                        .await
+                        .context("reopen logs because of exceeded size")?;
+                }
+            }
+
+            // Write the timestmap
+            self.file_mut().write_all(timestamp.as_bytes()).await?;
+
+            // Add the pipe name
+            match pipe {
+                Pipe::StdOut => self.file_mut().write_all(b" stdout ").await,
+                Pipe::StdErr => self.file_mut().write_all(b" stderr ").await,
+            }?;
+
+            // Output log tag for partial or newline
+            if partial {
+                self.file_mut().write_all(b"P ").await?;
+            } else {
+                self.file_mut().write_all(b"F ").await?;
+            }
+
+            // Output the actual contents
+            self.file_mut().write_all(&line_buf).await?;
+
+            // Output a newline for partial
+            if partial {
+                self.file_mut().write_all(b"\n").await?;
+            }
+
+            bytes_written += bytes_to_be_written;
+            trace!("Wrote log line of length {}", bytes_to_be_written);
+        }
+
+        Ok(())
+    }
+
+    /// Reopen the container log file.
+    pub async fn reopen(&mut self) -> Result<()> {
+        debug!("Reopen container log {}", self.path().display());
+        self.file().get_ref().sync_all().await?;
+        self.set_file(Self::open(self.path()).await?);
```

Comment:
I think we need to drop self.file first, so the Self::open doesn't open the already opened file.

Reply:
We cannot drop self.file because then we would partially move content out of `self`, which is only borrowed as mutable reference.

```diff
diff --git a/conmon-rs/server/src/cri_logger.rs b/conmon-rs/server/src/cri_logger.rs
index e34da9d7..b891b4fe 100644
--- a/conmon-rs/server/src/cri_logger.rs
+++ b/conmon-rs/server/src/cri_logger.rs
@@ -128,6 +128,7 @@ impl CriLogger {
     pub async fn reopen(&mut self) -> Result<()> {
         debug!("Reopen container log {}", self.path().display());
         self.file().get_ref().sync_all().await?;
+        drop(self.file);
         self.set_file(Self::open(self.path()).await?);
         Ok(())
     }
```

```
error[E0507]: cannot move out of `self.file` which is behind a mutable reference
   --> conmon-rs/server/src/cri_logger.rs:131:14
    |
131 |         drop(self.file);
    |              ^^^^^^^^^ move occurs because `self.file` has type `tokio::io::BufWriter<tokio::fs::File>`, which does not implement the `Copy` trait
```

`self.file` gets automatically dropped on replacement because it runs out of scope. For example:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7d454dac7b772a4e5c05e98a2f383e21

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5d77116cd57ac89e845ddb1844b2a368

This example opens the file, and then drops the reference.

Reply:
Perrhaps https://docs.rs/io-close/latest/io_close/

Reply:
I think for this scenario you need to make it an `Option<File>`.  Then it's just:
```
self.file = None;
self.set_file(...)
```

Reply:
Perhaps my concern here isn't warranted. The kubelet does the rotation and removes the old file, so the open is opening a completely new file.

Reply:
I'd agree with that.

Reply:
resolving... since the kubelet removes the file and then requests for a new one to be opened.